### ISSUE TITLE
Windows' WinFile is setting the time in the future

### ISSUE THREAD

**Issue by robert-j** [dosemu2/fdpp#222]
**Describe the bug**

Either WinFile or Windows itself is setting the time in the future (about 2 - 5 years. The amount is random).

**To Reproduce**

- Start WfW 3.11
- Start WinFile, browse the Windows folder
- exit Windows
- enter 'date'

**A regression?**

Yes. It started to happen a few days ago.

**Boot.log**

[date-bug.log.zip](https://github.com/dosemu2/dosemu2/files/11402027/date-bug.log.zip)

The file is actually **xz**-compressed.

**Comment by stsp** [comment]
```
struct dosdate
{
  unsigned short year;
  unsigned char monthday, month;
};
```
I don't see the packing problems
here (members are sorted by size),
and sizeof() gives 4.

**Comment by andrewbird** [comment]
Sizes look okay here, but I wondered if the version of clang used would affect things?

**Comment by stsp** [comment]
No because sizes are sorted properly:
"short" goes before smaller "char".
You may want to read about alignment
and padding rules, they are not essentially
trivial. :)

**Comment by andrewbird** [comment]
> You may want to read about alignment
and padding rules, they are not essentially
trivial. :)

Yes I've been a bit lazy on that front. I have it boiled down to just two options:
I don't care about alignment and padding as the struct is internal, so let the compiler do whatever,
and I do care because the struct has to be portable, so pack it.

**Comment by stsp** [comment]
That approach doesn't work if you
use `-Wpacking` option. It will warn
on redundant packing, which, on
clang-16, is not redundant at all!
So what clang-16 does, contradicts
itself. Of course, after I reported that,
they decided to break also `-Wpacking`
option, instead of fixing things up.
Because the "bug" is not following
gcc these days, rather than what
actually breaks things. :(

**Comment by stsp** [comment]
There are more problems to that
approach. Namely, whatever is packed,
is treated by gcc as misaligned! Even
when its actually not. So fdpp won't
even compile if you pack too much.
And unfortunately clang devs almost
threaten me with porting that bug from
gcc too. :(

**Comment by stsp** [comment]
So overall you can't pack w/o actually
reading about alignment and padding
rules.
If you do, you will either get a punishment
from `-Wpacking`, or, in case it is gcc -
will drown in its bugs, but your code won't
work.

**Comment by robert-j** [comment]
Is there a DPMI layer on top of the real mode INT 21 handler in dosemu? I'm asking because INT 21 2A/2B need a real mode to protected mode translation. Maybe is that place the culprit.

**Comment by stsp** [comment]
> Is there a DPMI layer on top of the real mode INT 21 handler in dosemu?

Yes.

> I'm asking because INT 21 2A/2B need a real mode to protected mode translation.

Not above the one mandated by dpmi
specs, where the register values are
propagated to PM as-is.

**Comment by robert-j** [comment]
I've tried this:

- start Windows+WinFile
- terminate Windows
- check date: yes, it's wrong
- write to a DOS-native file system, mounted as a floppy
- got this crash:

`ERROR: fdpp: abort at /build/fdpp-xbqHwh/fdpp-1.6-1352/fdpp/farptr.hpp:571`

This is 100% reproducible.

- when I start Windows and don't execute WinFile: no crash.
- when I run any other app in place of WinFile (Office, Delphi, Borland Pascal): no crash.

**Comment by stsp** [comment]
Please attach the log of the crash.

**Comment by robert-j** [comment]
[bug-222.log.xz.zip](https://github.com/dosemu2/fdpp/files/11426279/bug-222.log.xz.zip), actually an xz file.

I've redacted everything containing /drive_c/ for privacy reasons.

**Comment by stsp** [comment]
Please don't use -s when doing this log.
Also you need to install the dbgsym package
for fdpp and dosemu2. See this comment on how to:
https://github.com/dosemu2/dosemu2/issues/1987#issuecomment-1501099547

You need to grep the resulting log for
"No debugging symbols found"
and
"Could not attach to process."
strings. Once they are not there, the log
should be good.

**Comment by stsp** [comment]
Hmm, it seems fdpp compiler got confused
with this statement at fatfs.c:408:
`j = sp->sftt_count, sftp = sp->sftt_table`.
Its too difficult for him. :)
Could you please try this change:
```
diff --git a/kernel/fatfs.c b/kernel/fatfs.c
index 79498de..5ddec1e 100644
--- a/kernel/fatfs.c
+++ b/kernel/fatfs.c
@@ -405,7 +405,8 @@ STATIC int merge_file_changes(f_node_ptr fnp, int collect)
   i = 0;
   for (sp = sfthead; sp != (sfttbl FAR *) - 1; sp = sp->sftt_next)
   {
-    for(j = sp->sftt_count, sftp = sp->sftt_table; --j >= 0; sftp++, i++)
+    sftp = sp->sftt_table;
+    for(j = sp->sftt_count; --j >= 0; sftp++, i++)
     {
       if (i != fnp->f_sft_idx && sftp->sft_count != 0
           && fnp->f_dpb == sftp->sft_dcb
```
That would reduce the complexity and
our dumb compiler will probably understand.

**Comment by robert-j** [comment]
How do I configure dosemu2 to use my custom fdpp from /usr/local?

**Comment by stsp** [comment]
Could you please provide a ready to
reproduce hd image?

**Comment by robert-j** [comment]
Not an hd image but a "drive_c" together with the image file on which it crashes when I try to write on it after Windows terminates. Is this okay?

**Comment by stsp** [comment]
If fd image alone is enough (I have
win3.1 here too), then that alone.

**Comment by robert-j** [comment]
The image could be a new, empty one. I've just created a new one and it seems that the bug is not related to my image at all.

My .dosemurc is just this:

```
$_cpu_vm = "kvm"
$_floppy_a = "dos622.img"
$_attrs_support = (off)
```

dos622.img is the empty (but formatted) image.

**Comment by robert-j** [comment]
I've sent you an e-mail with the subject "FDPP Bug 222".

No, it's much easier:

- start Windows
- start WinFile
- close WinFile
- exit Windows

Then:

```
cd a:
dir > test
```

**Comment by stsp** [comment]
OK, it appeared to have nothing to
do with the compilation process. Its
a genuine freedos bug.

**Comment by robert-j** [comment]
@stsp, thank you!

The actual issue (date in the future) of this bug entry isn't fixed, though. You can reproduce the same way as the SFT bug:

- start Windows
- start WinFile
- exit Windows
- type `date` at the command prompt when back in DOS.

The SFT issue arose while we were arguing about the date bug.

**Comment by stsp** [comment]
So date shows the correct date and time.
What should I see?

**Comment by stsp** [comment]
> I've sent you an e-mail with the subject "FDPP Bug 222"

The URL from mail shows error.

**Comment by robert-j** [comment]
Yes, I have deleted the link when the bug was closed ;) I've sent you a new one.

You should see a date in the future: 2025-2028. I believe the time remains correct, although I did not explicitly check it.

**Comment by stsp** [comment]
Will try later.

**Comment by stsp** [comment]
Should now be fixed, good catch!

**Comment by stsp** [comment]
Also we recently fixed many other
windows regressions, and I suppose
windows support is in a good shape
again. :)
